radio/src/gui/colorlcd/theme_manager.cpp

```diff
@@ -21,27 +21,119 @@
+{
+  if ((strncmp(val, RGBSTRING, strlen(RGBSTRING)) == 0) && (val[val_len - 1] == ')')) {
+    int r, g, b;
+    int numTokens = sscanf(val, "RGB(%d,%d,%d)", &r, &g, &b);
+
+    if (numTokens == 3)
+      return RGB(r, g, b);
+
+  } else if (val_len > 2 && val[0] == '0' && (val[1] == 'x' || val[1] == 'X')) {
+    val += 2; val_len -= 2;
+
+    auto rgb24 = yaml_hex2uint(val, val_len);
+    return RGB((rgb24 & 0xFF0000) >> 16, (rgb24 & 0xFF00) >> 8, rgb24 & 0xFF);
+  }
+
+  TRACE("Theme: Invalid color value");
+  return 0;
+}
+
+static bool w_color(const YamlNode* node, uint32_t val, yaml_writer_func wf,
+                    void* opaque)
+{
+  uint32_t color = (uint32_t)GET_RED(val) << 16 |
+                   (uint32_t)GET_GREEN(val) << 8 |
+                   (uint32_t)GET_BLUE(val);
+
+  if (!wf(opaque, "0x", 2)) return false;
+  return wf(opaque, yaml_rgb2hex(color), 3 * 2);
+}
+
+PACK(struct YAMLThemeSummary
+{
+  char name[NAME_LENGTH + 1];
+  char author[AUTHOR_LENGTH + 1];
+  char info[INFO_LENGTH+1];
+});
+
+PACK(struct YAMLThemeColors
+{
+  uint32_t colors[COLOR_COUNT - 2];     // We don't use DEFAULT or CUSTOM
+});
+
+PACK(struct YAMLTheme
+{
+  struct YAMLThemeSummary summary;
+  struct YAMLThemeColors colors;
+  
+  YAMLTheme() {
+    memset(this, 0, sizeof(YAMLTheme));
+  }
+});
+
+static const struct YamlNode struct_YAMLThemeSummary[] = {
+  YAML_STRING("name", (NAME_LENGTH + 1)),
+  YAML_STRING("author", (AUTHOR_LENGTH + 1)),
+  YAML_STRING("info", (INFO_LENGTH + 1)),
+  YAML_END
+};
+
+static const struct YamlNode struct_YAMLThemeColors[] = {
+  YAML_UNSIGNED_CUST( "PRIMARY1", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "PRIMARY2", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "PRIMARY3", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "SECONDARY1", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "SECONDARY2", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "SECONDARY3", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "FOCUS", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "EDIT", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "ACTIVE", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "WARNING", 32, r_color, w_color ),
+  YAML_UNSIGNED_CUST( "DISABLED", 32, r_color, w_color ),
+  YAML_END
+};
+
+static const struct YamlNode struct_YAMLTheme[] = {
+  YAML_STRUCT("summary", (NAME_LENGTH + AUTHOR_LENGTH + INFO_LENGTH + 3) * 8, struct_YAMLThemeSummary, NULL),
+  YAML_STRUCT("colors", (COLOR_COUNT - 2) * 32, struct_YAMLThemeColors, NULL),
+  YAML_END
+};
+
+static const struct YamlNode themeRootNode = YAML_ROOT(struct_YAMLTheme);
+
 static const char *colorNames[COLOR_COUNT] = {
     "DEFAULT",    "PRIMARY1",   "PRIMARY2",   "PRIMARY3",
     "SECONDARY1", "SECONDARY2", "SECONDARY3", "FOCUS",
     "EDIT",       "ACTIVE",     "WARNING",    "DISABLED", "CUSTOM",
 };
 
-#define HEX_COLOR_VALUE_LEN 8
-constexpr const char *RGBSTRING = "RGB(";
 constexpr const char* SELECTED_THEME_FILE = THEMES_PATH "/selectedtheme.txt";
```

Comment:
If there is any way you can see to finally getting rid of this also, and instead relying on the `themeName` parameter in `RADIO/radio.yml` that would be great. As it would finally bring all the configuration setting into yaml proper, and opening up avenues for future work. Both the radio firmware and companion look at that file to determine the active theme.

Reply:
> If there is any way you can see to finally getting rid of this also, and instead relying on the `themeName` parameter in `RADIO/radio.yml` that would be great. As it would finally bring all the configuration setting into yaml proper, and opening up avenues for future work. Both the radio firmware and companion look at that file to determine the active theme.

The 'themeName' property in the radio.yml files appears to be meant for some other use related to the OpenTXTheme class. The radio writes an empty string to this field while Companion does not write it at all to radio.yml.

Can it be repurposed as the user selected theme (field size will need to be increased)?

Reply:
I can't actually see the purpose of that themeName property any more - this is all very old OTX -> ETX code, so has never been cleaned up properly. i.e. g_eeGeneral.themeName is only used in two places, and the radio actually just writes `themeName: ""` whether you are using the default theme or a custom theme... :shrug: I think it's also the reason we still have a `FlyskyTheme` knocking around...